runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/BatchFlinkExecutableStageContext.java

```diff
@@ -57,7 +57,8 @@ public StateRequestHandler getStateRequestHandler(
     return FlinkBatchStateRequestHandler.forStage(executableStage, runtimeContext);
   }
 
-  private void cleanUp() throws Exception {
+  @Override
+  public void finalize() throws Exception {
```

Comment:
In original line [77](https://github.com/apache/beam/pull/5692/files#diff-6e80fb44d46373766f4b831f10a9fcc1L77), removal.getValue() is null as value is already garbage collected which throws NPE and stops cleanup from ever happening.
Doing cleanup in finalize of the BatchFlinkExecutableStageContext will keep the intended behavior of closing the jobBundleFactory when BatchFlinkExecutableStageContext goes out of scope.

Reply:
This should not be the case. We are not using weak values directly. They are instead managed by the cache for us. I tested this by hand to make sure  clean up is called and the original way appeared to work as expected. Have you seen this null pointer exception being thrown in practice?

In general, it is preferable to use explicit clean up by cache (as in the original) over finalizers.

Reply:
Yes, there was null pointer exceptions which led me to investigate this. 
The cache only wraps and unwraps the value in in weak reference.
The remove Listener called on cache maintenance which happens at a later time by which the value is already garbage collected.

I agree that using finalize something which is not good but in this case it does the intended behavior and is no worst than weakReference. Both weakReference and finalize depends upon the garbage collector's behavior.